python/cuml/testing/strategies.py

```diff
@@ -12,13 +12,127 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+import cudf
+import cupy as cp
+import numpy as np
+from cuml.common.array import CumlArray
 from hypothesis import assume
-from hypothesis.extra.numpy import arrays, floating_dtypes
-from hypothesis.strategies import composite, integers, just, none, one_of
+from hypothesis.extra.numpy import arrays, floating_dtypes, array_shapes
+from hypothesis.strategies import (composite, integers, just, none, one_of,
+                                   sampled_from)
+from numba import cuda
 from sklearn.datasets import make_regression
 from sklearn.model_selection import train_test_split
 
 
+_CUML_ARRAY_INPUT_TYPES = [
+    'numpy', 'numba', 'cupy', 'series', None,
+]
+
+
+_CUML_ARRAY_DTYPES = [
+    np.float16, np.float32, np.float64,
+    np.int8, np.int16, np.int32, np.int64,
+    np.uint8, np.uint16, np.uint32, np.uint64
+]
+
+_CUML_ARRAY_ORDERS = ["F", "C"]
+
+
+_CUML_ARRAY_SUPPORTED_OUTPUT_DTYPES = [
+    np.float16, np.float32, np.float64,
+    np.int8, np.int16, np.int32, np.int64,
+    np.uint8, np.uint16, np.uint32, np.uint64
+]
+
+
+_UNSUPPORTED_CUDF_DTYPES = [np.uint8, np.uint16, np.uint32, np.uint64,
+                            np.float16]  # TODO: maybe not needed?
+
+
+@composite
+def cuml_array_input_types(draw):
+    return draw(sampled_from(_CUML_ARRAY_INPUT_TYPES))
+
+
+@composite
+def cuml_array_dtypes(draw):
+    return draw(sampled_from(_CUML_ARRAY_DTYPES))
+
+
+@composite
+def cuml_array_orders(draw):
+    return draw(sampled_from(_CUML_ARRAY_ORDERS))
+
+
+@composite
+def cuml_array_shapes(
+    draw,
+    *,
+    min_dims=1,
+    max_dims=2,
+    min_side=1,
+    max_side=None
+):
+    max_side = 10 if max_side is None else max_side
+
+    # TODO: Raise ValueError for these:
+    assert 1 <= min_dims <= max_dims
+    assert 0 < min_side < max_side
+
+    shapes = array_shapes(
+        min_dims=min_dims, max_dims=max_dims,
+        min_side=min_side, max_side=max_side,
+    )
+    return draw(shapes)
+
+    # just_size = integers(min_side, max_side)
+    # return draw(one_of(shapes, just_size))
+
+
+def _create_cuml_array_input(input_type, dtype, shape, order):
+    multidimensional = isinstance(shape, tuple) and \
+        len([d for d in shape if d > 1]) > 1
+    assume(
+        not (
+            input_type == "series"
+            and (
+                dtype in _UNSUPPORTED_CUDF_DTYPES
+                or multidimensional
+            )
+        )
+    )
+
+    array = cp.ones(shape, dtype=dtype, order=order)
+
+    if input_type == 'numpy':
+        return np.array(cp.asnumpy(array), dtype=dtype, order=order)
+
+    elif input_type == 'numba':
+        return cuda.as_cuda_array(array)
```

Comment:
The explicit `numba` array is a legacy inheritance from before the cuda_array_interface days. With CAI, numba can be treated as one more CAI supporting arrray type, so we should remove the explicit testing of it.

Reply:
Got it. For back-wards compatible we should still test for `to_output("numba")`, or not?

Reply:
I'd prefer to leave the numba tests in for now, just to make sure #4908 doesn't have some weird breaking change, but I agree that deprecating a lot of our numba infrastructure immediately after that makes good sense.